moveit_ros/moveit_servo/src/utils/common.cpp

```diff
@@ -0,0 +1,333 @@
+  joint_trajectory.header.stamp = rclcpp::Time(0);
+  joint_trajectory.header.frame_id = servo_params.planning_frame;
+  joint_trajectory.joint_names = joint_state.joint_names;
+
+  trajectory_msgs::msg::JointTrajectoryPoint point;
+  point.time_from_start = rclcpp::Duration::from_seconds(servo_params.publish_period);
+
+  // Set the fields of trajectory point based on which fields are requested.
+  // Some controllers check that acceleration data is non-empty, even if accelerations are not used
+  // Send all zeros (joint_state.accelerations is a vector of all zeros).
+  point.positions = servo_params.publish_joint_positions ? joint_state.positions : point.positions;
+  point.velocities = servo_params.publish_joint_velocities ? joint_state.velocities : point.velocities;
+  point.accelerations = servo_params.publish_joint_accelerations ? joint_state.accelerations : point.accelerations;
+
+  joint_trajectory.points.push_back(point);
+  return joint_trajectory;
+}
+
+std::pair<double, StatusCode> velocityScalingFactorForSingularity(const moveit::core::RobotStatePtr& robot_state,
+                                                                  const Eigen::VectorXd& target_delta_x,
+                                                                  const servo::Params& servo_params)
+{
+  // We need to send information back about if we are halting, moving away or towards the singularity.
+  StatusCode servo_status = StatusCode::NO_WARNING;
+
+  const moveit::core::JointModelGroup* joint_model_group =
+      robot_state->getJointModelGroup(servo_params.move_group_name);
+
+  // Get the thresholds.
+  const double lower_singularity_threshold = servo_params.lower_singularity_threshold;
+  const double hard_stop_singularity_threshold = servo_params.hard_stop_singularity_threshold;
+  const double leaving_singularity_threshold_multiplier = servo_params.leaving_singularity_threshold_multiplier;
+
+  // Get size of total controllable dimensions.
+  size_t dims = target_delta_x.size();
+
+  // Get the current Jacobian and compute SVD
+  Eigen::JacobiSVD<Eigen::MatrixXd> current_svd = Eigen::JacobiSVD<Eigen::MatrixXd>(
+      robot_state->getJacobian(joint_model_group), Eigen::ComputeThinU | Eigen::ComputeThinV);
+  Eigen::MatrixXd matrix_s = current_svd.singularValues().asDiagonal();
+
+  // Compute pseudo inverse
+  Eigen::MatrixXd pseudo_inverse = current_svd.matrixV() * matrix_s.inverse() * current_svd.matrixU().transpose();
+
+  // Get the singular vector corresponding to least singular value.
+  // This vector represents the least responsive dimension. By convention this is the last column of the matrix U.
+  // The sign of the singular vector from result of SVD is not reliable, so we need to do extra checking to make sure of
+  // the sign. See R. Bro, "Resolving the Sign Ambiguity in the Singular Value Decomposition".
+  Eigen::VectorXd vector_towards_singularity = current_svd.matrixU().col(dims - 1);
```

Comment:
const?

Reply:
This cannot be const as it could be updated with a sign flip later.